Module(cash): Implement the cash payments module

jniles opened the pull request:
This PR re-implements the client-side cash payments module.  The cash payments module now supports the payment of either Caution Payments (against future invoices) or Invoice Payments (against past invoices).

**Notable features, changes, additions**
1. [FIX] Critical bug fixes in the cash API.  Additional integration tests added to ensure the API behaves appropriately when information is missing.
2. Receipts are displayed in a modal, rather than a page `$location` change.
3. Added [ngMessages](https://docs.angularjs.org/api/ngMessages/directive/ngMessages) for awesome form validation.  Updated Angular as a result. See https://github.com/IMA-WorldHealth/bhima-2.X/commit/675e1a718a8da032a41d08184743a97ae874a0b0.
4. [FIX] Critical bug fixes in the ExchangeRateService.  The code has been dramatically improved for clarity and readabilty.  Additional service methods have also been implemented.  See https://github.com/IMA-WorldHealth/bhima-2.X/commit/d5231210da812a76eef776194bb87aa812982780 and https://github.com/IMA-WorldHealth/bhima-2.X/commit/90d8f2ffe259cfaa9dec2853055a8cb59f20ec8a.
5. [FEAT] UI Grid now has a generic loading, error, and no data messages.  See https://github.com/jniles/bhima-2.X/commit/d5231210da812a76eef776194bb87aa812982780.
6. [FEAT] Massive improvements to test helpers!  The test helpers now enforces API standards by requiring users to send back the correct HTTP codes and error codes.  These make the process of writing integration tests **a lot** easier.  See https://github.com/jniles/bhima-2.X/commit/d5231210da812a76eef776194bb87aa812982780 for more details.
7. Updated jshintrc rules to remove deprecated or removed rules from the jshint tool.  See https://github.com/jniles/bhima-2.X/commit/2f1b953b046aa3b44a84dd19e3f95c0b819a20d2.

You should read the commit messages for the details of changes - they are well summarized.

jniles committed:
CHORE(util): clean up util files

This commit cleans up the utilities file by adding in detailed comments
and standardizing variable names.  Ideally, the file should be removed
completely in favor of a tested NPM module (see #32).

Additionally, the /cash route has been uncommented in the database files
and build-db.sh has been updated to no longer call schema-v2.sql

jniles committed:
CHORE(cash): implemented client-side form

This commit re-writes the client-side of the cash module.

jniles committed:
CHORE(bower): install angular-messages 1.4.9

This commit introduces a dependency on angular-messages.  Requires `bower
install` to be run in the local repository (potentially `bower update` as
well).

jniles committed:
CHORE(ngMessage): template in generic messages

This commit introduces a shared template for messages to be used in
forms all across bhima, dependent on ngMessage.  These include messages
for common issues like "required", "min", "max", and others.

jniles committed:
WIP(cash): moved cashbox selection out of /cash

This commit removes cashbox selection from the /cash controller and puts
it in its own controller/template.  This greatly reduces the complexity
in the cash controller, but requires a better solution for state.

The routes now look like this:
 - /cash matches to the cashbox selection page
 - /cash/:id attempts to load a cashbox from the server matching :id.
   If none exists, reroutes to the /cash page.

In the future, this state-like behavior should be encapsulated in
something like ui-router.

jniles committed:
WIP(cash): work in progress on cash module

This commit separates out cash modals into their own files and
implements the patient find directive directly on the page.
Furthermore, the interface for switching cashboxes has been cleaned up
considerably.

jniles committed:
FEAT(cash): caution payment now possible

This commit finalizes the proceedure for creating a caution payment.
The following features have been implemented in support of this:
 1. The form now locks when submitting, to prevent double clicks on the
 submit button. A loading indicator appears in the button as well.
 This behavior should change once #57 lands.
 2. The cash API correctly handles date parsing.

jniles committed:
CHORE(cash): Allow patient invoice selections

This commit adds a ui-grid modal to the cash module.

jniles committed:
CHORE(misc): Improvements to cash dependencies

This commit provides a wide array of improvements to the dependencies of
the cash payments module.
 1. The ExchangeRateService's methods been tested properly to return the
 correct results for convertToEnterpriseCurrency() and
 convertFromEnterpriseCurrency().
 2. The ExchangeRateService now includes getCurrentRate(id) for a given
 currency id.
 3. Added a utility class .pad-bottom to make sure columns do not run
 together when stacking on small screens.
 4. Documentation improvements to CurrencyService, ExchangeRateService,
 and others.
 5. Removed poorly conceived CSS rules.
 6. Added CSS classes to support loading indicators in ui-grid's css.
 TODO: migrate these to LESS.
 7. Fixed a bug in the debtors API, resulting in a crash when looking up
 debtor invoices that had never made a purchase.

jniles committed:
FEAT(cash): cash payments against invoices

This commit adds the ability to submit payments against the cash API.
It includes usability improvements with the cash invoice modal,
including the ability to remember prior selections of invoices in the
modal.

jniles committed:
CHORE(cash): improvements to API testing

This commit builds tooling around the cash API tests.  In particular,
test helpers have been implemented to respond specifically to the bhima
API requests, allowing tests to test logic instead of boiler plate.

Specifically:
  - [Bug] CashService converts is_caution to a Number as necessary.
  - [Optimization]: Error interceptors no longer return httpStatus.
  - [Feat] helpers.js includes all boilerplate API specific validation.
  - [Bug] Cash API parses multiple invoice items correctly
  - [Bug] Reject cash invoice payments without invoice items.

jniles committed:
CHORE(cash): receipt modeal code, misc fixes

This commit does the following:
  1. Remove httpStatus from client-sent errors
  2. Add in openReceiptModal into cash controller
  3. initial work on Receipt Modal for cash

jniles committed:
CHORE(cash): partially display receipts in modal

This commit fixes a few bugs and adds the ability to view cash receipts
in a modal.  The complete functionality with Exchange Rates and others
have not been completed fully.

jniles committed:
CHORE(cash): Finalize receipt rendering

This commit fixes the cash receipt rendering. The following changes have
been made:
  1. ExchangeRateService simplified to use the same code to lookup a
  previous exchange rate.  Huge improvement to maintainability and
  testability.
  2. English translations added.  French translations are a TODO.
  3. The PatientService has been brought up to 2.X standards.  Notably
    - uses util.unwrapHttpResponse()
    - instantiated by angular.service
  4. Cleaner layout for receipt modal HTML
  5. Loading indicator while data loads.

Feature Request: Currencies.round(currencyId, amount) to round
currencies to their nearest min monentary unit.  Since we store values
in a single currency, we need a way to reliably represent real-world
payments when converting from stored values.  Currently, it is psosible
for the cash receipt to read "102.45FC", which obviously is not correct.

jniles committed:
CHORE(docs): cleaned up cash documentation

This commit cleans up the cash module's documenation and syntax to
standard.  Many TODOs have been documented to reflect issues brought up
on Github.

jniles committed:
CHORE(lint): rm deprecated jshint properties

This commit brings our .jshintrc up to date with the latest jshintrc.
For reference, see
https://github.com/jshint/jshint/blob/master/examples/.jshintrc  Notable
changes inlcude:
 - enable mocha and nodejs environments.
 - remove all environments (rhino, wsh, etc) that don't make sense for
   us
 - "predef" --> "globals"

New enforcements:
 - require Object.hasOwnProperty() for `for ... in` loops.
 - disallow bitwise operations unless specifically escaped

jniles committed:
TEST(cash): implement e2e tests for cashbox select

This commit implements a variety of tests for the cashbox selection
process that takes place before the cash module is able to be used.

sfount commented on client/src/partials/cash/cash.html:
```diff
@@ -1,137 +1,217 @@
+                  <label>{{ "COLUMNS.PAYMENT" | translate }}</label>
+                  <div class="input-group">
+                    <span class="input-group-addon">
+                      {{ CashCtrl.payment.currency_id ? CashCtrl.currencies.symbol(CashCtrl.payment.currency_id) : '...' }}
```
This comment will not block the pull request. This method will be called on every angular digest, meaning if any unrelated input changes the whole statement will be re-evaluated (calling the function x times, often up to 10 times). 

Is it possible to move this calculation into a variable that is updated when the currency_id changes? The optimisation is very small in this case however it is considered good practice.

jniles replied:
:+1: will do.

sfount commented on client/src/partials/templates/messages.tmpl.html:
```diff
@@ -0,0 +1,23 @@
+<!-- displays "This value is too small for this field." -->
+<p ng-message="min">This value is too small for this field.</p>
```
This is a great separation of concerns. These messages should eventually use translation keys.

jniles replied:
:+1:  Would you like me to include translations in this PR?

sfount replied:
If it will not take very long. There's no reason at this stage in 
development it should be blocking.

On 09/02/2016 11:55, Jonathan Niles wrote:

> In client/src/partials/templates/messages.tmpl.html 
> https://github.com/IMA-WorldHealth/bhima-2.X/pull/89#discussion_r52292848:
> 
> > @@ -0,0 +1,23 @@
> > +<!-- displays "This value is too small for this field." -->
> > +<p ng-message="min">This value is too small for this field.</p>
> 
> :+1: Would you like me to include translations in this PR?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/IMA-WorldHealth/bhima-2.X/pull/89/files#r52292848.

sfount commented on client/src/partials/cash/modals/receipt.modal.js:
```diff
@@ -0,0 +1,118 @@
+      // bind the receipt
+      vm.receipt = receipt;
+
+      return $q.all([
```
It has been discussed that this many HTTP requests could be well grouped on the server and then sent to the service in one request. This works for now however just a comment to remind about the API design.

jniles committed:
TEST(cash): implement end to end test for caution

This commit implements an end to end test to demonstrate that cautions
work.  The test should be refactored as more components get added to
use.

jniles committed:
TEST(cash): e2e test for invoice payment

This commit introduces a single test to demonstrate that
payments against previous invoices work without a hitch.  It also
includes the first demonstration of ui-grid test utilities in bhima.

jniles committed:
CHORE(cash): view rendering optimizations

This commit implements the optimizations proposed by @sfount in the cash
module.

sfount commented on client/src/partials/cash/cash.html:
```diff
@@ -1,137 +1,219 @@
+                  <label>{{ "COLUMNS.PAYMENT" | translate }}</label>
+                  <div class="input-group">
+                    <span class="input-group-addon">
+                      {{ CashCtrl.currencyLabel }}
```
This is exactly what I was thinking :+1: 

@IMA-WorldHealth/local-contributors This is an excellent example of replacing a function call with a variable that is calculated in the controller - reducing the number of times Angular has to check the value of the method call.
